### PR TITLE
[HOTFIX] Fix app template webhooks

### PR DESCRIPTION
**Pull request by krasish**

### CODE REVIEW

File: components/director/internal/domain/apptemplate/converter_test.go [hunk 1]
```diff
@@ -301,6 +301,88 @@ func TestConverter_InputFromGraphQL(t *testing.T) {
...
+				conv := &automock.WebhookConverter{}
+				conv.On("MultipleInputFromGraphQL", []*graphql.WebhookInput(nil)).Return([]*model.WebhookInput(nil), nil)
+				return conv
+			},
+			Input:         graphql.ApplicationTemplateUpdateInput{},
+			Expected:      model.ApplicationTemplateUpdateInput{},
+			ExpectedError: nil,
+		},
+		{
+			Name: "Error when converting",
```
Comment (KirilKabakchiev) [inline components/director/internal/domain/apptemplate/converter_test.go:349]: add one more case for the second converter error as well

File: components/director/internal/domain/apptemplate/service_test.go [hunk 2]
```diff
@@ -419,18 +419,18 @@ func TestService_List(t *testing.T) {
 func TestService_Update(t *testing.T) {
 	// GIVEN
 	ctx := tenant.SaveToContext(context.TODO(), testTenant, testExternalTenant)
-	modelAppTemplate := fixModelApplicationTemplate(testID, testName, []*model.Webhook{})
+	modelAppTemplate := fixModelApplicationTemplate(testID, testName, nil)
```
Comment (KirilKabakchiev) [inline components/director/internal/domain/apptemplate/service_test.go:422]: any particular reason for making this nil ?

File: components/director/internal/domain/webhook/repository.go [hunk 3]
```diff
@@ -146,7 +162,7 @@ func (r *repository) Update(ctx context.Context, item *model.Webhook) error {
 	if err != nil {
 		return errors.Wrap(err, "while converting model to entity")
 	}
-	return r.updater.UpdateSingle(ctx, entity)
+	return r.updaterGlobal.UpdateSingleGlobal(ctx, entity)
```
Comment (KirilKabakchiev) [inline components/director/internal/domain/webhook/repository.go:165]: It looks to me that now i can get a client cert for app A in tenant 1 and update a webhook via the graphql api for app B in tenant 2 (security issue)

File: components/director/internal/domain/webhook/service.go [hunk 4]
```diff
@@ -48,11 +49,7 @@ func NewService(repo WebhookRepository, appRepo ApplicationRepository, uidSvc UI
 }
 
 func (s *service) Get(ctx context.Context, id string) (*model.Webhook, error) {
-	tnt, err := tenant.LoadFromContext(ctx)
-	if err != nil {
-		return nil, err
-	}
-	webhook, err := s.webhookRepo.GetByID(ctx, tnt, id)
+	webhook, err := s.webhookRepo.GetByIDGlobal(ctx, id)
```
Comment (KirilKabakchiev) [inline components/director/internal/domain/webhook/service.go:52]: with this now i can delete a webhook for app that belongs to a tenant different from mine (similar to the update security issue above)

File: components/director/pkg/graphql/app_template_validation.go [hunk 5]
```diff
@@ -69,6 +69,62 @@ func (i ApplicationTemplateInput) ensurePlaceholdersUsed() error {
...
+	if err := i.ensureUniquePlaceholders(); err != nil {
+		return err
+	}
+	if err := i.ensurePlaceholdersUsed(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (i ApplicationTemplateUpdateInput) ensureUniquePlaceholders() error {
```
Comment (KirilKabakchiev) [inline components/director/pkg/graphql/app_template_validation.go:92]: several methods here are present twice in the same file.. thats too much duplication

File: tests/director/api/application_templates_api_test.go [hunk 6]
```diff
@@ -194,3 +195,63 @@ func TestRegisterApplicationFromTemplate(t *testing.T) {
 	require.Equal(t, "test new-value", *outputApp.Application.Description)
 	saveExample(t, createAppFromTmplRequest.Query(), "register application from template")
 }
+
+func TestUpdateApplicationTemplateWebhooks(t *testing.T) {
```
Comment (KirilKabakchiev) [inline tests/director/api/application_templates_api_test.go:199]: Its more like TestAddWebhookToApplicationTemplate - as you are using addWebhook mutation and not updateApplicationTemplate here

File: tests/director/api/application_templates_api_test.go [hunk 7]
```diff
@@ -194,3 +195,63 @@ func TestRegisterApplicationFromTemplate(t *testing.T) {
...
+
+	// add
+	url := "http://new-webhook.url"
+	urlUpdated := "http://updated-webhook.url"
+	webhookInStr, err := tc.graphqlizer.WebhookInputToGQL(&graphql.WebhookInput{
+		URL:  &url,
+		Type: graphql.WebhookTypeUnregisterApplication,
+	})
+
+	require.NoError(t, err)
```
Comment (KirilKabakchiev) [inline tests/director/api/application_templates_api_test.go:216]: t.log here as well ? like you did before the create application template step

File: tests/director/api/application_templates_api_test.go [hunk 8]
```diff
@@ -194,3 +195,63 @@ func TestRegisterApplicationFromTemplate(t *testing.T) {
...
+	err = tc.RunOperation(ctx, addReq, &actualWebhook)
+	require.NoError(t, err)
+	assert.NotNil(t, actualWebhook.URL)
+	assert.Equal(t, "http://new-webhook.url", *actualWebhook.URL)
+	assert.Equal(t, graphql.WebhookTypeUnregisterApplication, actualWebhook.Type)
+	id := actualWebhook.ID
+	require.NotNil(t, id)
+
+	// get all webhooks
+	updatedAppTemplate := getApplicationTemplate(t, ctx, appTemplate.ID)
```
Comment (KirilKabakchiev) [inline tests/director/api/application_templates_api_test.go:230]: t.log ?

File: tests/director/api/application_templates_api_test.go [hunk 9]
```diff
@@ -194,3 +195,63 @@ func TestRegisterApplicationFromTemplate(t *testing.T) {
...
+	assert.Len(t, updatedAppTemplate.Webhooks, 1)
+
+	// update
```
Comment (KirilKabakchiev) [inline tests/director/api/application_templates_api_test.go:233]: t.log is more useful then these comments, add that one before each graphql call

File: tests/director/api/application_templates_api_test.go [hunk 10]
```diff
@@ -194,3 +195,63 @@ func TestRegisterApplicationFromTemplate(t *testing.T) {
...
+	updateReq := fixUpdateWebhookRequest(actualWebhook.ID, webhookInStr)
+	err = tc.RunOperation(ctx, updateReq, &actualWebhook)
+	require.NoError(t, err)
+	assert.NotNil(t, actualWebhook.URL)
+	assert.Equal(t, urlUpdated, *actualWebhook.URL)
+
+	// delete
+
+	//GIVEN
+	deleteReq := fixDeleteWebhookRequest(actualWebhook.ID)
```
Comment (KirilKabakchiev) [inline tests/director/api/application_templates_api_test.go:248]: you could do one more get call to verify its no longer present

**Review by KirilKabakchiev — CHANGES_REQUESTED** [review]

File: components/director/internal/domain/apptemplate/converter_test.go (hunk 1 above)
Comment (krasish) [reply components/director/internal/domain/apptemplate/converter_test.go:349]: Actually, all webhook converters will be removed, bc applicaitonUpdateInput no does not support updating. Will do that in next commit.

File: components/director/internal/domain/apptemplate/service_test.go (hunk 2 above)
Comment (krasish) [reply components/director/internal/domain/apptemplate/service_test.go:422]: Update in service layer accepts ApplicaitonTemplateUpdateInput which has no webhooks in it. Consequently, the app template passed to the repo layer will contain no Webhooks. Webhooks are updated using the add/update/deleteWebhook mutations.

File: components/director/internal/domain/webhook/repository.go (hunk 3 above)
Comment (krasish) [reply components/director/internal/domain/webhook/repository.go:165]: Fixed.

**Pull request closed without merging**